### ISSUE TITLE
Inconsistent behaviour on `collection=objects` and `collection_singular_ids=ids` for `has_many :through` association

### ISSUE THREAD

**Issue by febeling** [rails/rails#33942]
### Steps to reproduce

```
order.items = [item1, item1]
order.items.count # will be 2
```
while
```
order.items = [item1]
order.items = [item1, item1]
order.items.count # now will be 1
```

See this gist https://gist.github.com/febeling/06e1da8ad34d5a9c45d41420efbe58bc for the full test cases.

### Expected behavior
With has_many-through and join-models duplicates are technically possibly, as different from regular has_many associations. And sometimes that is a useful tool.

When I assign the whole new array of objects or ids to be associated, these should be the ones associated - no matter if they where associated before or not. (Or raise an error if duplicates were illegal.)

### Actual behavior
There is inconsistent behaviour on assignment of object arrays or also id arrays that contain duplicates. The outcome depends on whether an object is _already_ present in the association when its added again, or not.

It it is already present, it will not be added again.

If it's not already present, it will be added multiple times - even with a `replace` operation, which explicitly discards prior state.

### System configuration
**Rails version**:
5.1-6.0.0.alpha I have confirmed to show the bug
**Ruby version**:
2.4.5

**Comment by febeling** [comment]
These places below seem to assume a set character of the collection, which would break with duplicates. But I'm not sure I understand the inheritance hierarchy correctly, yet.

https://github.com/rails/rails/blob/fc5dd0b85189811062c85520fd70de8389b55aeb/activerecord/lib/active_record/associations/collection_association.rb#L412  
https://github.com/rails/rails/blob/fc5dd0b85189811062c85520fd70de8389b55aeb/activerecord/lib/active_record/associations/collection_association.rb#L414  
https://github.com/rails/rails/blob/fc5dd0b85189811062c85520fd70de8389b55aeb/activerecord/lib/active_record/associations/collection_association.rb#L64

**Comment by febeling** [comment]
I think the difference `-` and union `&` operations don't work correctly, neither does the `size` comparison on L64, with `:through` assocs, if they contains duplicates. They would need to consider identity as established by the join model instead of the joined model on the far end of the association.

**Comment by darwin67** [comment]
I confirmed the inconsistent behaviour is still happening on `master` and I agree the default behaviour with no additional options should be at least consistent.

Do you have a preference of a behaviour over the other? @febeling 
I personally think that it would be better to allow duplication since you can always add validations to make in more strict in the model level or in the database level.

**Comment by febeling** [comment]
@darwin67 Yes, I'd prefer allowing duplicates. I think many other users likely are relying on them as well. So ruling them out would not be a smooth change, in my estimation. 

I have been thinking about the problem though, and believe to have a reasonable solution for it. The idea is to treat each occurrence of a record as an identifiable or unique element (occurence). The can established just by counting and keeping score in a temporary hash. Decorating the record collections this way for the set operations mentioned in an earlier comment might be enough. 

If that works out I'll prepare a PR.
